validator/validator_spawner.go

```diff
@@ -3,33 +3,191 @@ package validator
 import (
 	"context"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"os"
 	"path/filepath"
+	"runtime"
+	"sync/atomic"
 	"time"
 
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	flag "github.com/spf13/pflag"
+
 	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/log"
-	"github.com/pkg/errors"
 )
 
+type ReadyMarker interface {
+	Ready() bool
+	ReadyChan() chan struct{}
+	WaitReady(ctx context.Context) error
+}
+
 type ValidationSpawner interface {
-	Execute(ctx context.Context, entry *ValidationInput, moduleRoot common.Hash) (GoGlobalState, error)
+	Launch(entry *ValidationInput, moduleRoot common.Hash) ValidationRun
+	Start(context.Context)
 	Stop()
 	Name() string
+	Room() int
+}
+
+type ValidationRun interface {
+	ReadyMarker
+	WasmModuleRoot() common.Hash
+	Result() (GoGlobalState, error)
+	Close()
+}
+
+type ArbitratorSpawnerConfig struct {
+	ConcurrentRuns     int    `koanf:"concurrent-runs-limit" reload:"hot"`
+	OutputPath         string `koanf:"output-path" reload:"hot"`
+	TargetMachineCount int    `koanf:"target-machine-count"`
+}
+
+type ArbitratorSpawnerConfigFecher func() *ArbitratorSpawnerConfig
+
+var DefaultArbitratorSpawnerConfig = ArbitratorSpawnerConfig{
+	ConcurrentRuns:     0,
+	OutputPath:         "./target/output",
+	TargetMachineCount: 4,
+}
+
+func ArbitratorSpawnerConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.Int(prefix+".concurrent-runs-limit", DefaultArbitratorSpawnerConfig.ConcurrentRuns, "number of cuncurrent runs")
+	f.String(prefix+".output-path", DefaultArbitratorSpawnerConfig.OutputPath, "path to write machines to")
+	f.Int(prefix+".target-machine-count", DefaultArbitratorSpawnerConfig.TargetMachineCount, "target machine count")
```

Comment:
Let's rename `target-machine-count` to `cached-challenge-machines` and `concurrent-runs-limit` to `parallel-validation-threads`. The current names have confused pretty much everyone who used them.

Reply:
I noticed that TargetMachineCount was moved in the next PR, so you might want to base this change on top of the final PR. Feel free to do that for all of my recommendations btw.

Reply:
I did option renames on the last pr in the series.
Instead of parallel-validation-threads It's now "validation.[arbitrator/jit].workers" LMK if you like that name or not.